app/maptools/recordingmaptool.cpp

```diff
@@ -336,11 +349,41 @@ bool RecordingMapTool::hasValidGeometry() const
     }
     else if ( mLayer->geometryType() == QgsWkbTypes::LineGeometry )
     {
-      return mRecordedGeometry.constGet()->nCoordinates() >= 2;
+      if ( mRecordedGeometry.isMultipart() )
+      {
+        const QgsAbstractGeometry *geom = mRecordedGeometry.constGet();
+        for ( auto it = geom->const_parts_begin(); it != geom->const_parts_end(); ++it )
+        {
+          if ( ( *it )->nCoordinates() < 2 )
+          {
+            return false;
+          }
+        }
+        return true;
+      }
+      else
+      {
+        return mRecordedGeometry.constGet()->nCoordinates() >= 2;
+      }
     }
     else if ( mLayer->geometryType() == QgsWkbTypes::PolygonGeometry )
     {
-      return mRecordedGeometry.constGet()->nCoordinates() >= 3;
+      if ( mRecordedGeometry.isMultipart() )
+      {
+        const QgsAbstractGeometry *geom = mRecordedGeometry.constGet();
+        for ( auto it = geom->const_parts_begin(); it != geom->const_parts_end(); ++it )
+        {
+          if ( ( *it )->nCoordinates() < 3 )
+          {
+            return false;
+          }
+        }
+        return true;
+      }
+      else
+      {
+        return mRecordedGeometry.constGet()->nCoordinates() >= 3;
```

Comment:
Shouldn't it actually be 4? (When ring is closed, with three points it is 4 in total - first duplicated)

```suggestion
        return mRecordedGeometry.constGet()->nCoordinates() >= 4;
```